### ISSUE TITLE
TAB completion on REPL makes identifier confliction

### ISSUE THREAD

**Issue by appleparan** [JuliaLang/julia#23677]
This is from [FFTW.jl issue #36](https://github.com/JuliaMath/FFTW.jl/issues/36#issuecomment-328506720)

TAB completion on REPL, especially using `using` statement for module, seems loading identifier once again. 

As I said on [this](https://github.com/JuliaMath/FFTW.jl/issues/36#issuecomment-328851240), **using TAB completion** even `using Base` makes `FFTW` identifier confliction error. 
```
root@67c40e951d84:~# julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: https://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.7.0-DEV.1568 (2017-08-29 23:34 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 881ab49 (13 days old master)
|__/                   |  x86_64-pc-linux-gnu

julia> using Base

julia> using FFTW
WARNING: error while reinitializing value FFTW:
ErrorException("cannot assign variable DFT.FFTW from module Main")
WARNING: requiring "FFTW" in module "Main" did not define a corresponding module.

julia>
```

I think [this line of `deprecated.jl`](https://github.com/JuliaMath/FFTW.jl/issues/36#issuecomment-328851240) makes Julia load the identifier again. Do we need to change deprecated macro usage or change TAB completion?

**Comment by stevengj** [comment]
I'm guessing that this is a problem with tab completion and how it [searches through modules](https://github.com/JuliaLang/julia/blob/780407c36726ee8625048cb27a4cd98f9421a021/base/repl/REPLCompletions.jl#L64-L80).

**Comment by appleparan** [comment]
Yeah, FFTW symbols should not be searched on REPL tab completion via [`Base.deprecated`](https://github.com/JuliaLang/julia/blob/f5691695bc1a5d0cbc49249a8915ad44ce7f717b/base/reflection.jl#L95), but it is included because of [`deprecated.jl`](https://github.com/JuliaLang/julia/blob/fce0a3c9ebf05d38d62c821ba83d461051fbfd7b/base/deprecated.jl#L1422)

**Comment by stevengj** [comment]
Maybe if we call `Base.deprecate(Base, :DFT, 2)` and update the completion code somewhere to look for the deprecation flag...?

**Comment by stevengj** [comment]
A non-interactive way to get the warning:
```jl
julia> Base.REPLCompletions.completions("using Base", 10)
(["Base"], 7:10, true)

julia> using FFTW
WARNING: error while reinitializing value FFTW:
ErrorException("cannot assign variable DFT.FFTW from module Main")
WARNING: requiring "FFTW" in module "Main" did not define a corresponding module.
```

**Comment by ararslan** [comment]
It's tempting, now that packages have had a chance to adopt FFTW.jl, to remove the structure I put in place in base/deprecated.jl that preserves the `Base.DFT.FFTW` module hierarchy. That was so that code calling into the module (e.g. to access FFT flags) wouldn't be suddenly broken. We could keep the deprecations for the symbols that were exported from Base, e.g. `fft` and `ifft`, at the top level and instead add `FFTW` and `DFT` to the special cases in [the `showerror` method for `UndefVarError`](https://github.com/JuliaLang/julia/blob/02fba855595f5f4ea1dc12ec3fe1cff4926ef959/base/replutil.jl#L282).

**Comment by stevengj** [comment]
If you remove it, then code will be suddenly broken when people upgrade to 0.7, which seems to be contrary to our usual deprecation practice.

In any case, it seems like it would be good to fix this.  I don't think `DFT` will be the last module that gets moved out of Base.

**Comment by JeffBezanson** [comment]
Maybe fixed by #29962?

**Comment by KristofferC** [comment]
Let's assume it is and reopen / new issue if it comes up again.
